src/python/pants/engine/legacy/graph.py

```diff
@@ -296,6 +389,58 @@ class HydratedTargets(Collection.of(HydratedTarget)):
   """An intransitive set of HydratedTarget objects."""
 
 
+class OwnersRequest(datatype([
+  ('sources', tuple),
+  ('include_dependees', str),
+])):
+  """A request for the owners (and optionally, transitive dependees) of a set of file paths."""
+
+
+def find_owners(symbol_table, address_mapper, owners_request):
+  sources_set = set(owners_request.sources)
+  dirs_set = {dirname(source) for source in sources_set}
+  # Walk up the buildroot looking for targets that would conceivably claim changed sources.
+  candidate_specs = tuple(AscendantAddresses(directory=d) for d in dirs_set)
+
+  def owns_any_source(legacy_target):
+    """Given a `HydratedTarget` instance, check if it owns the given source file."""
+    target_kwargs = legacy_target.adaptor.kwargs()
+
+    # Handle `sources`-declaring targets.
+    # NB: Deleted files can only be matched against the 'filespec' (ie, `PathGlobs`) for a target,
+    # so we don't actually call `fileset.matches` here.
+    target_sources = target_kwargs.get('sources', None)
+    if target_sources and any_matches_filespec(sources_set, target_sources.filespec):
+      return True
+
+    return False
+
+  # Request HydratedTargets for our candidate specs, and then match their source globs against
+  # the sources set.
+  hydrated_targets = yield Get(HydratedTargets, Specs(candidate_specs))
+  direct_owners = tuple(ht.adaptor.address
+                        for ht in hydrated_targets
+                        if LegacyAddressMapper.any_is_declaring_file(ht.adaptor.address, sources_set) or
+                           owns_any_source(ht))
+
+  # If the OwnersRequest does not require dependees, then we're done: otherwise, find dependees.
+  if owners_request.include_dependees not in ('direct', 'transitive'):
+    # NB: A yield without a LHS represents an early return.
```

Comment:
Because this is non-obvious, I'd prefer the rest of the function to be in an `else` block (which is very against my ordinary code style - boo Python!)

Reply:
An alternative is:
```python
yield last
return  # Never reached
```

But I think I agree with @illicitonion here with the added benefit of the `assert owners_request.include_dependees == 'transitive'` probably being un-needed at that point.

Reply:
It's not possible to mix `yield` and `return`... I'll nest the remainder of the body.

Reply:
It is when `return` returns nothing, it ends the generator. Right now the generator is only early return because the controlling code in `nodes.rs` stops its loop when it receives its 1st non Get*. It could continue to advance the generator in this code and the python generator machinery would not stop early but continue on to the Gets.

For example:
```python
>>> def coroutine(exit_early):
...   if exit_early:
...     yield 42
...   yield 37
...   x = yield
...   y = yield
...   yield x * y
...   yield 1/137
... 
>>> cr_early = coroutine(True)
>>> next(cr_early)
42
>>> next(cr_early)
37
>>> next(cr_early)
>>> cr_early.send(3)
>>> cr_early.send(5)
15
>>> next(cr_early)
0
>>> next(cr_early)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
StopIteration
```

Versus an early return forced by the controlled code:
```python
>>> def coroutine(exit_early):
...   if exit_early:
...     yield 42
...     return
...   yield 37
...   x = yield
...   y = yield
...   yield x * y
...   yield 1/137
... 
>>> cr_early = coroutine(True)
>>> next(cr_early)
42
>>> next(cr_early)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
StopIteration
```

Versus the syntax error you're thinking of:
```python
>>> def coroutine(exit_early):
...   if exit_early:
...     yield 42
...     return 3
...   yield 37
...   x = yield
...   y = yield
...   yield x * y
...   yield 1/137
... 
  File "<stdin>", line 4
SyntaxError: 'return' with argument inside generator
```

Reply:
Ahhh. You are completely correct. Sorry.